R/telemetry.R

```diff
@@ -0,0 +1,587 @@
+      checkmate::assert_string(id)
+
+      logger::log_debug("click: {id}", namespace = "shiny.telemetry")
+
+      private$.log_event(
+        event = "click",
+        id = id,
+        session = session
+      )
+    },
+
+    #' @description
+    #' Log the browser version
+    #'
+    #' @param session ShinySession object or NULL to identify the current
+    #' Shiny session.
+
+    log_browser_version = function(
+      session = shiny::getDefaultReactiveDomain()
+    ) {
+
+      input <- shiny::reactiveValues()
+      if (checkmate::test_r6(session, "ShinySession")) {
+        input <- session$input
+      }
+
+      shiny::observeEvent(input$browser_version, {
+        browser <- input$browser_version
+        shiny::validate(
+          shiny::need(browser, "'browser_info_js' should be set in app head")
+        )
+        logger::log_debug(
+          "browser_version: {browser}", namespace = "shiny.telemetry"
+        )
+
+        private$.log_event(
+          event = "browser",
+          value = browser,
+          session = session
+        )
+      })
+    },
+
+    #' @description
+    #' Log session details
+    #'
+    #' @param detail string with details about the session.
+    #' @param session ShinySession object or NULL to identify the current
+    #' Shiny session.
+
+    log_session = function(
+      detail, session = shiny::getDefaultReactiveDomain()
+    ) {
+      checkmate::assert_string(detail)
+
+      logger::log_debug("session_details: {detail}", namespace = "shiny.telemetry")
+
+      private$.log_session(
+        detail = detail,
+        session = session
+      )
+    },
+
+    #' @description
+    #' Track a button and track changes to this input (without storing the
+    #' values)
+    #'
+    #' @param input_id string that identifies the button in the Shiny
+    #' application so that the function can track and log changes to it.
+    #' @param session ShinySession object or NULL to identify the current
+    #' Shiny session.
+
+    log_button = function(
+      input_id, session = shiny::getDefaultReactiveDomain()
+    ) {
+      self$log_input(input_id, session = session)
+    },
+
+    #' @description
+    #' A short description...
+    #'
+    #' @param track_values flag that indicates if the basic telemetry should
+    #' track the values of the inputs that are changing. `FALSE` by default.
+    #' This parameter is ignored if `track_inputs` is `FALSE`.
+    #' @param excluded_inputs vector of input_ids that should not be tracked.
+    #' By default it doesn't track browser version, which is added by this
+    #' package.
+    #' @param session ShinySession object or NULL to identify the current
+    #' Shiny session.
+
+    log_all_inputs = function(
+      track_values = FALSE,
+      excluded_inputs = c("browser_version"),
+      session = shiny::getDefaultReactiveDomain()
+    ) {
+      input <- shiny::reactiveValues()
+      if (checkmate::test_r6(session, "ShinySession")) {
```

Comment:
When can `session` be something different than `ShinySession` instance?

Reply:
Some weird scenario I was sure that existed, but cannot remember right now.